sklearn/utils/stats.py

```diff
@@ -40,33 +46,41 @@ def _weighted_percentile(array, sample_weight, percentile=50):
     # When sample_weight 1D, repeat for each array.shape[1]
     if array.shape != sample_weight.shape and array.shape[0] == sample_weight.shape[0]:
         sample_weight = np.tile(sample_weight, (array.shape[1], 1)).T
+
+    # Sort `array` and `sample_weight` along axis=0; set `sample_weight` for nan input
+    # to 0:
     sorted_idx = np.argsort(array, axis=0)
     sorted_weights = np.take_along_axis(sample_weight, sorted_idx, axis=0)
+    sorted_nan_mask = np.take_along_axis(np.isnan(array), sorted_idx, axis=0)
+    sorted_weights[sorted_nan_mask] = 0
```

Comment:
> Is it more efficient?

It might be a bit more efficient to skip a memory allocation of size `(n_samples, n_features)` for the result of for `sorted_nan_mask`.

Note that `np.isnan(array)` would also lead to an extra allocation of size `(n_samples, n_features)`. So it would also be great to avoid it when there are no nans.

> Also, in case we have this condition, would it be better be if np.isnan(array[-1, ...]).any()

I believe so, but I don't understand the `-1` indexing, why would the last row be any special? `array` is not sorted at this point. I believe the intention was:


```python
# May contain NaN, which would sort to the end. If there are any, let's set
# their weight to zero but we do so under an if condition to avoid temporary
# allocations of size `(n_samples, n_features)` when there are no NaN values
# in the input data array.
n_features = array.shape[1]
largest_value_per_column = array[sorted_idx[-1, ...], np.arange(n_features)]
if np.isnan(largest_value_per_column).any():
    ...
```

`largest_value_per_column` has size `(n_features,)` so its allocation should not matter and we can protect against the other two `(n_samples, n_features)` allocations under the `if` block.